### ISSUE TITLE
[🐞] Cloudflare Pages: Ability to run cloudflare pages dev server with vite

### ISSUE THREAD

**Issue by sujith97** [QwikDev/qwik-evolution#307]
### Which component is affected?

Qwik Rollup / Vite plugin

### Describe the bug

The current Cloudflare pages setup makes it hard to develop qwik-city with Cloudflare page functions. Cloudflare has D1, KV etc objects that should be accessible during the development mode but at the moment we should first build. The build drops the static files to `dist` folder which can then be accessed using Cloudflare pages. Ideally, we should be able to have the HMR and wrangler reload during local development.

Can we provide guidance on how to develop locally with Cloudflare and qwik-city vite? (https://developers.cloudflare.com/pages/platform/functions/local-development/)

```
export const useListLoader = routeLoader$(async (context: any) => {
  const ps = (context.env.get("cloudflaredb") as any)?.prepare(
    "SELECT * FROM users"
  );
  if (ps) {
    const users = await ps.first();
    return {
      users,
    };
  } else {
    console.log("No access to D1", context.env.get("cloudflaredb"));
  }

  return {
    users: [],
  };
});
```

### Reproduction

https://stackblitz.com/edit/qwik-starter-joopk8?file=src%2Froutes%2Findex.tsx

### System Info

```shell
System:
    OS: macOS 13.0
    CPU: (8) arm64 Apple M1 Pro
    Memory: 279.02 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.15.0 - ~/.volta/tools/image/node/18.15.0/bin/node
    Yarn: 1.22.19 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
    npm: 9.5.0 - ~/.volta/tools/image/node/18.15.0/bin/npm
  Browsers:
    Chrome: 111.0.5563.64
    Safari: 16.1
  npmPackages:
    @builder.io/qwik: 0.20.1 => 0.20.1
    @builder.io/qwik-city: 0.5.2 => 0.5.2
    undici: 5.20.0 => 5.20.0
    vite: 4.1.4 => 4.1.4
```

**Comment by flikteoh** [comment]
Hello,

Has this issue been addressed, or will it be addressed in V2?

Currently, Cloudflare offers [Pages](https://developers.cloudflare.com/pages/framework-guides/deploy-a-qwik-site/) and [Workers with Assets](https://developers.cloudflare.com/workers/frameworks/framework-guides/qwik/) that supports Qwik Framework.

According to their [blog post](https://blog.cloudflare.com/blazing-fast-development-with-full-stack-frameworks-and-cloudflare/), we would be able to start up the [local development with access to service bindings](https://developers.cloudflare.com/workers/testing/local-development/) (KV, D1, R2).

However, this is currently not the case when running `npm run dev`. Qwik works in SSR in dev mode and does not generate any files, and thus getting the error:

```
my-qwik-app✨npx wrangler dev

 ⛅️ wrangler 3.87.0
-------------------

✘ [ERROR] The entry-point file at "dist/_worker.js" was not found.

```

This happens in both Pages and Workers with Assets.

Having access to bindings locally can give faster feedback loops and better DX.

I think currently [SvelteKit](https://svelte.dev/docs/kit/adapter-cloudflare#Options-platformProxy), [Astro](https://docs.astro.build/en/guides/integrations-guide/cloudflare/#platformproxy), [Remix](https://remix.run/docs/en/main/guides/vite#cloudflare-proxy) has implemented this via Wrangler's API: [getPlatformProxy](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy).

It would be nice if this can also be supported in Qwik, as currently the only way to work with bindings is to deploy to Cloudflare on every iterations.

Here's their [Workers vs Pages compatibility matrix](https://developers.cloudflare.com/workers/static-assets/compatibility-matrix/) for references.

Thank you

**Comment by wmertens** [comment]
I'm thinking this can be resolved with https://developers.cloudflare.com/workers/vite-plugin/ and the current v2 branch?

there are a couple more tweaks we can do to the plugin to fully use environments but I think you can already use a workerd environment.

**Comment by flikteoh** [comment]
> I'm thinking this can be resolved with https://developers.cloudflare.com/workers/vite-plugin/ and the current v2 branch?
> 
> there are a couple more tweaks we can do to the plugin to fully use environments but I think you can already use a workerd environment.

Yes that will definitely be more ideal. Their vite plugin currently only support a few frameworks officially, which needs other framework maintainers to help with integration.

I'm currently working on (trying) Cloudflare Worker adapter for Workers with Static Asset, instead of trying to make changes to the existing Cloudflare Pages Adapter since I'm not sure how much it will affect existing usage. Currently the [`C3 create cloudflare` cli](https://github.com/cloudflare/workers-sdk/tree/main/packages/create-cloudflare/templates/qwik) is [still using Pages adapter](https://developers.cloudflare.com/workers/framework-guides/web-apps/more-web-frameworks/qwik/). 

Not quite sure if the maintainers will accept another Cloudflare Adapter in this case. My thinking is to create a new adapter for the Workers so setting up Qwik on Workers can be more straightforward and less confusing. Easier setup hopefully means better adoptions.

I've seen many similar questions and issues in their discord channel when people are [switching from Pages to Workers](https://developers.cloudflare.com/workers/static-assets/migration-guides/migrate-from-pages/) with Static Assets because there are some similarities and differences during setup.

Existing Pages adapter doesn't seem to [support usage](https://github.com/cloudflare/workers-sdk/blob/b1fe9d7af62780468492162347802cb8ebfdce28/packages/create-cloudflare/templates/qwik/workers/c3.ts#L111) like `cf: { ... }` or I could be wrong?

**Comment by wmertens** [comment]
I know that CF has been working on making workerd work with Qwik, but I'm not sure where

**Comment by flikteoh** [comment]
They were only maintaining the [C3 cli template thingy](https://github.com/cloudflare/workers-sdk/tree/main/packages/create-cloudflare/templates/qwik) for all framework I guess. Nothing was done on workerd itself since that is only their worker runtime?

I assume what we need is either to create a worker adapter or update the pages adapter. But seeing that they are putting Pages in maintenance only mode, I guess I should be focusing on creating a worker adapter similar with our current Pages adapter but support worker's features instead? Am I on the right path?

Also: https://github.com/cloudflare/workers-sdk/blob/main/packages/create-cloudflare/templates/qwik/workers/snippets/getPlatformProxy.ts

**Comment by wmertens** [comment]
I'm moving this to qwik-evolution since it's a feature request.

Right now the v2 Router vite setup is much simplified, it only adds a middleware that runs the equivalent of the preview server, and it uses a stream transformation to inject dev helpers into the SSR results.

So two things stand in the way for this to work:
1. fully support environment API in qwikVite, meaning to work in a way that allows running both client and ssr builds at the same time, looking up info inside hooks instead of statically outside of the hooks
2. come up with a common API for letting adapters run in dev mode. That's a different RFC.
